### PR TITLE
Bugfix/force compile

### PR DESCRIPTION
**Pull request by mitzimorris**
#### Submission Checklist

- [x] Run unit tests
- [x] Declare copyright holder and open-source license: see below

#### Summary

Cleanup of `compile` command logic with unit test to check that compile arg `force` does the right thing.

Also add logic to CmdStanModel ctor and `sample` method to handle CmdStan https://github.com/stan-dev/cmdstan/issues/1057

#### Copyright and Licensing

Please list the copyright holder for the work you are submitting (this will be you or your assignee, such as a university or company):  Columbia University

By submitting this pull request, the copyright holder is agreeing to license the submitted work under the following licenses:
- Code: BSD 3-clause (https://opensource.org/licenses/BSD-3-Clause)

### CODE REVIEW

**Comment by mitzimorris** [comment]
still needs more unit tests, including interaction w/ `fixed_param` and `num_chains`

File: test/test_model.py [hunk 1]
```diff
@@ -153,22 +149,31 @@ def test_cpp_options(self):
...
+        # model.compile(force=True, cpp_options=override_opts)
+        # dict2 = model.exe_info
+        # get_logger().info('dict 2 %s', dict2)
+        # self.assertEqual(dict2['STAN_THREADS'].lower(), 'false')
+
+        model.compile(
+            force=True, cpp_options=override_opts, override_options=True
+        )
+        dict3 = model.exe_info
+        get_logger().info('dict 3 %s', dict3)
```
Comment (WardBrian) [inline test/test_model.py:175]: Why are we logging in our tests (as opposed to `print`)?

File: test/test_model.py [hunk 2]
```diff
@@ -153,22 +149,31 @@ def test_cpp_options(self):
...
     def test_model_info(self):
-        # copy so that parallel compile tests don't mess up exe
-        b2_filename = os.path.join(DATAFILES_PATH, 'b2.stan')
-        b2_file = shutil.copyfile(BERN_STAN, b2_filename)
-        cpp_opts = {'STAN_THREADS': 'TRUE'}
-        model = CmdStanModel(stan_file=b2_file, cpp_options=cpp_opts)
-        if model.exe_file is not None and os.path.exists(model.exe_file):
-            os.remove(model.exe_file)
-        empty_dict = model.exe_info()
-        self.assertEqual(len(empty_dict), 0)
```
Comment (WardBrian) [inline test/test_model.py:164]: I think this is a reasonable test to still have

File: cmdstanpy/model.py [hunk 3]
```diff
@@ -942,17 +942,19 @@ def sample(
...
             one_process_per_chain = True
             assert isinstance(self.exe_file, str)  # make typechecker happy
-            info_dict = self.exe_info()
+            info_dict = self.exe_info
             stan_threads = info_dict.get('STAN_THREADS', 'false').lower()
+            # run multi-chain sampler if possible, unless algo is fixed_param
             if (
                 force_one_process_per_chain is None
                 and not cmdstan_version_before(2, 28, info_dict)
+                and not fixed_param
```
Comment (WardBrian) [inline cmdstanpy/model.py:951]: Should probably just use `force_one_process_per_chain = force_one_process_per_chain or fixed_param`

**Review by WardBrian — COMMENTED** [review]
Comments for pair programming

**Comment by WardBrian** [comment]
I'll do some more Windows-specific bug hunting tomorrow

**Comment by WardBrian** [comment]
Just a thought: We will now recompile if:

1. `force` is set to True
2. The compiler options are different.

Would it make sense to add "if the existing exe file was compiled with an older version of cmdstan than the current interpreter is using"? This would be useful but essentially impossible to test in our CI

**Comment by WardBrian** [comment]
@mitzimorris We need some tests that instantiate a CmdStanModel class with the exe file only. Just caught a bug in testing where model._fixed_param was not initialized in this case, so .sample failed

**Comment by WardBrian** [comment]
An extension I think we should allow: in the CmdStanModel constructor, we currently allow you to set `compile=True|False`. I should you should also be able to set it `compile="force"` which does the obvious thing.

If you think this is a good idea @mitzimorris I'll implement it tomorrow before doing the Windows debugging

**Comment by mitzimorris** [comment]
> Would it make sense to add "if the existing exe file was compiled with an older version of cmdstan than the current interpreter is using"? This would be useful but essentially impossible to test in our CI

I'm not sure - why don't you ask on Discourse?  once a model is written and working, new language features don't matter, so the only benefit would be bugfixes/performance enhancements in Stan and/or mathlib.  which sometimes aren't as great as hoped.

**Comment by mitzimorris** [comment]
> in general I think we should actually call sample etc on models constructed this way

not sure what you mean here.

the problem is that compilation takes many seconds if not a minute.  consider the teaching/demo use case:  there's a lecture's worth of model and data plus notebook.  new CmdStan release comes out, so user upgrades.  then goes to teach from the notebook - suddenly everything's recompiling - possibly unexpected - kinda ruins the packing of the show.

**Comment by WardBrian** [comment]
Oh my last comment was related to the unit test you just added. I think it’s important we call .sample(), .variational() and .optimize() on CmdStanModel objects which are exe only in the unit tests, not just construct them

**Comment by mitzimorris** [comment]
just pushed changes to fix logic around fixed_param and exe only.  will add more unit tests today.

**Comment by WardBrian** [comment]
Could you explain the idea behind having the fixed_param flag be None by default? The logic doesn't seem to use it anywhere specifically

**Comment by WardBrian** [comment]
Found issue: TemporaryCopiedFile returns a sanitized path even if `is_copied` is False. We should use this unconditionally as we have before. Doing that and some other clean up now, including renaming it to `SanitziedOrTmpFilePath`

**Comment by mitzimorris** [comment]
w/r/t fixed_param = None,  will check the logic again - one of the added unit tests triggered it.

**Comment by WardBrian** [comment]
As far as I can tell, the logic we have right now is completely ambivalent as to whether fixed_param is None or False, so we can just make it false. I'm in the middle of some logic refactors so I'd hold off for now to avoid merge issues

**Comment by WardBrian** [comment]
Seems to now be passing on all platforms. Will look into `compile="force"` behavior and general cleanup.

**Comment by WardBrian** [comment]
The solution I found in general required passing `shell=True` to `Popen` on Windows. This is generally not recommended, but without it the spec says Popen may completely ignore both `env` and `cwd` arguments:

>Resolving the path of executable (or the first item of args) is platform dependent. For POSIX, see os.execvpe(), and note that when resolving or searching for the executable path, cwd overrides the current working directory and env can override the PATH environment variable. For Windows, see the documentation of the lpApplicationName and lpCommandLine parameters of WinAPI CreateProcess, and note that when resolving or searching for the executable path with shell=False, cwd does not override the current working directory and env cannot override the PATH environment variable. Using a full path avoids all of these variations.

**Comment by WardBrian** [comment]
If I come back from lunch and all tests pass then I think I'm happy with the state of this. That said, feel free to add more tests and I can them if they fail

**Review by WardBrian — COMMENTED** [review]
I think this counts as my final review, for whatever that means in the pair-programming system we used for this PR

File: docsrc/workflow.rst [hunk 4]
```diff
@@ -64,10 +67,6 @@ The constructor argument ``compile=False`` will override the default behavoir.
     my_model.exe_file
     my_model.code()
 
-The method :meth:`~CmdStanModel.compile` is used to compile the model as needed.
```
[the remark this answers is not in the document]
Comment (mitzimorris) [reply docsrc/workflow.rst:67]: workflow page is def a work in progress.  am trying to explain the overlap between constructor args and compile args.  I will continue this in a separate PR.

I can restore that sentence and then I think we should merge this PR.

[the remark this answers is not in the document]
Comment (WardBrian) [reply docsrc/workflow.rst:67]: I agree

**Comment by WardBrian** [comment]
We seem to be getting that intermitten `Operation not permitted` error in variational tests - I don't recall what the solution was the last time we did this, but I'm also unsure why it would be happening now? Seems to be the exe_only test

**Comment by mitzimorris** [comment]
> We seem to be getting that intermittent Operation not permitted error in variational tests - I don't recall what the solution was the last time we did this, but I'm also unsure why it would be happening now? 

will investigate - I think ADVI sometimes fails on that model unless you specify a specific seed, etc.  which is why all other calls in varatiational test have all args specified.  I forgot this detail when adding the new tests

**Comment by WardBrian** [comment]
Great. I'm also having an issue with test_utils.test_cmdstan_version clobbering my CMDSTAN env variable on windows and then causing all subsequent tests to fail, so I'd like to fix that before merge as well

**Comment by WardBrian** [comment]
I don't have all the dev tools set up properly on my windows machine, apologies for linting issues. If tests pass this is good to go IMO
